src/pybind/mgr/volumes/fs/operations/versions/subvolume_v2.py

```diff
@@ -0,0 +1,434 @@
+                st = self.fs.stat(subvol_path)
+
+                self.uid = int(st.st_uid)
+                self.gid = int(st.st_gid)
+                self.mode = int(st.st_mode & ~stat.S_IFMT(st.st_mode))
+        except MetadataMgrException as me:
+            if me.errno == -errno.ENOENT:
+                raise VolumeException(-errno.ENOENT, "subvolume '{0}' does not exist".format(self.subvolname))
+            raise VolumeException(me.args[0], me.args[1])
+        except cephfs.ObjectNotFound:
+            log.debug("missing subvolume path '{0}' for subvolume '{1}'".format(subvol_path, self.subvolname))
+            raise VolumeException(-errno.ENOENT, "mount path missing for subvolume '{0}'".format(self.subvolname))
+        except cephfs.Error as e:
+            raise VolumeException(-e.args[0], e.args[1])
+
+    def _get_clone_source(self):
+        try:
+            clone_source = {
+                'volume'   : self.metadata_mgr.get_option("source", "volume"),
+                'subvolume': self.metadata_mgr.get_option("source", "subvolume"),
+                'snapshot' : self.metadata_mgr.get_option("source", "snapshot"),
+            }
+
+            try:
+                clone_source["group"] = self.metadata_mgr.get_option("source", "group")
+            except MetadataMgrException as me:
+                if me.errno == -errno.ENOENT:
+                    pass
+                else:
+                    raise
+        except MetadataMgrException as me:
+            raise VolumeException(-errno.EINVAL, "error fetching subvolume metadata")
+        return clone_source
+
+    @property
+    def status(self):
+        state = SubvolumeStates.from_value(self.metadata_mgr.get_global_option(MetadataManager.GLOBAL_META_KEY_STATE))
+        subvolume_type = self.subvol_type
+        subvolume_status = {
+            'state' : state.value
+        }
+        if not SubvolumeOpSm.is_complete_state(state) and subvolume_type == SubvolumeTypes.TYPE_CLONE:
+            subvolume_status["source"] = self._get_clone_source()
+        return subvolume_status
+
+    @property
+    def state(self):
+        return SubvolumeStates.from_value(self.metadata_mgr.get_global_option(MetadataManager.GLOBAL_META_KEY_STATE))
+
+    @state.setter
+    def state(self, val):
+        state = val[0].value
+        flush = val[1]
+        self.metadata_mgr.update_global_section(MetadataManager.GLOBAL_META_KEY_STATE, state)
+        if flush:
+            self.metadata_mgr.flush()
+
+    @property
+    def type(self):
+        return SubvolumeTypes.from_value(self.metadata_mgr.get_global_option(MetadataManager.GLOBAL_META_KEY_TYPE))
+
+    def trash_incarnation_dir(self):
+        self._trash_dir(self.path)
+
+    def remove(self, retainsnaps=False):
+        if self.list_snapshots():
+            if not retainsnaps:
+                raise VolumeException(-errno.ENOTEMPTY, "subvolume '{0}' has snapshots".format(self.subvolname))
```

Comment:
This is correct errno when a removal of subvolume is issued which has snapshots and also the subvolume exists. 
But this error is also getting hit for subvolume which doesn't exist i.e.,for those subvolume which are deleted with "--retain-snapshots" option. Ideally those should be returned with ENOENT.

```
# bin/ceph fs subvolume rm a sub_0 --retain-snapshots
# 
# bin/ceph fs subvolume rm a sub_0 
Error ENOTEMPTY: subvolume 'sub_0' has snapshots
```

Reply:
If subvolume does not have a current <uuid> directory but has snapshots, it should return ENOTEMPTY, not ENOENT, right? The subvolume is not empty and hence cannot be deleted.

ENOENT would mean subvolume does not exist and further `remove_subvolume_snapshot` understands ENOENT as a success (idempotent remove being a success) and reports it as such to the user.

Reply:
I was coming from user's perspective. User might think subvolume and snapshots are different entities. If he has removed subvolume with retaining snapshots, the subvolume no longer exists for him. So I was thinking trying to remove the same subvolume should say "Error ENOENT: subvolume 'xyz' is removed and has only snapshots retained" along the lines of 'subvolume getpath'. 

But I think you made sense there, 'subvolume rm without retain snapshots' should say ENOTEMPTY and 'subvolume rm with retain-snapshot' could be idempotent or say ENOENT.

I am fine with the existing code.